emailproxy.py

```diff
@@ -337,9 +344,157 @@ def add_account(username):
+    def _aws_secrets_debug_log_boto3_error(err):
+        Log.debug(err.response['Error']['Code'] + ":",
+                  err.response['Error']['Message'])
+
+    @staticmethod
+    def _aws_secrets_fetch():
+        aws_client = AppConfig._aws_secrets_boto3_client()
+        if aws_client:
+            # get the values of all AWS Secrets referenced in the configuration file
+            aws_secrets = AppConfig._aws_secrets_ids()
+            for secret_id in aws_secrets:
+                try:
+                    get_secret_value_response = aws_client.get_secret_value(SecretId=secret_id)
+
+                except botocore.exceptions.ClientError as err_get_secret:
+                    if err_get_secret.response['Error']['Code'] == 'ResourceNotFoundException':
+                        Log.info('Fetching AWS Secret "%s"' % secret_id, '- secret does not exist or does not contain',
+                                 'a secret value')
+                    elif err_get_secret.response['Error']['Code'] == 'AccessDeniedException':
+                        Log.error('Fetching AWS Secret "%s"' % secret_id, '- access denied: does the IAM user have the',
+                                  '"secretsmanager:GetSecretValue" permission?')
+                    else:
+                        Log.error('Fetching AWS Secret "%s"' % secret_id, '- unexpected error; see the proxy\'s debug',
+                                  'mode log for details')
+                    AppConfig._aws_secrets_debug_log_boto3_error(err_get_secret)
+
+                else:
+                    aws_secrets[secret_id] = json.loads(get_secret_value_response['SecretString'])
+
+            # update local proxy configuration in memory
+            for account in AppConfig._aws_secrets_accounts():
+                account_secret = AppConfig._PARSER.get(account, 'aws_secret')
+                if account in aws_secrets[account_secret]:
+                    for key in AppConfig._SECRETS_MANAGER_KEYS:
+                        AppConfig._PARSER.set(account, key, aws_secrets[account_secret][account][key])
+        else:
+            Log.error('Unable to get AWS client; cannot fetch OAuth 2.0 tokens from AWS Secrets Manager')
+
+    @staticmethod
+    def _aws_secrets_save():
+        # a deep copy of the configuration file to be saved locally, *without* OAuth 2.0 tokens (_SECRETS_MANAGER_KEYS)
+        appconfig_to_save = configparser.ConfigParser()
+        appconfig_to_save.read_dict(AppConfig._PARSER)
+
+        # remove OAuth 2.0 tokens from each account, transferring them into a local dict of referenced AWS Secrets
+        aws_secrets = AppConfig._aws_secrets_ids()
+        for account in AppConfig._aws_secrets_accounts():
+            account_secret = appconfig_to_save.get(account, 'aws_secret')
+            aws_secrets[account_secret][account] = {key: appconfig_to_save.get(account, key) for key in
+                                                    AppConfig._SECRETS_MANAGER_KEYS}
```

Comment:
One thing we haven't covered so far is what would/should happen if the required keys are not present (either in the secret, or in the proxy's cache).

I don't know if/how this could happen from the AWS side, but the proxy can (and does) delete values when they become invalid, which would cause this method to fail.

One simple solution here would be to use a fallback value in `get()` and skip keys if they have no value (and entire accounts if there are no valid keys). What do you think?

Reply:
Another related question: my assumption is that token renewal is handled via a local copy of the proxy - i.e., you receive an authentication failure when using your remote server, which prompts you to open a local copy and authenticate, updating the AWS secret. Is that correct? If so, how do you trigger a token from AWS Secrets Manager on the remote server update?

Reply:
```suggestion
            aws_secrets[account_secret][account] = {key: appconfig_to_save.get(account, key, '') for key in
                                                    AppConfig._SECRETS_MANAGER_KEYS}
```

Reply:
I addressed the first question about required keys in a review below.

> Another related question: my assumption is that token renewal is handled via a local copy of the proxy - i.e., you receive an authentication failure when using your remote server, which prompts you to open a local copy and authenticate, updating the AWS secret. Is that correct? If so, how do you trigger a token update from AWS Secrets Manager on the remote server?

I think there are 3 options:

1. handle token renewal locally and use SIGHUP to reload config on the remote server.
2. handle token renewal locally and restart the remote server's proxy to reload config
3. handle the token renewal by ssh-ing into the remote server.

Reply:
I was wondering whether it is worth triggering `AppConfig._aws_secrets_fetch()` (probably via a more generic method in the AppConfig class) after a login failure for an account using AWS Secrets. This way, a login failure on the server would trigger checking for updated credentials that have been provided via a local login, and there would be no need to log in to the server at all.

Reply:
> I was wondering whether it is worth triggering AppConfig._aws_secrets_fetch() (probably via a more generic method in the AppConfig class) after a login failure for an account using AWS Secrets. This way, a login failure on the server would trigger checking for updated credentials that have been provided via a local login, and there would be no need to log in to the server at all.

@simonrob that's an excellent idea. Would it make sense to call `AppConfig._load()` to generically reload the config after a login failure? Thus reloading both the local and remote config.

Reply:
There's no actual need to reload the whole file, though that is a potential option (via `AppConfig.reload()`).

Either way, the main question is _when_ to actually do this. The trigger to set up a local copy and provide updated tokens is a login failure, so although [this](https://github.com/michaelstepner/email-oauth2-proxy/blob/e11338d444218fbd7f59b2a4f86d933066c1db24/emailproxy.py#L665) is a natural place, there's no way you could actually set up and authenticate via a local copy in the meantime between this failure and the subsequent no-cache retry.

A more viable option is to reload from AWS [here](https://github.com/michaelstepner/email-oauth2-proxy/blob/e11338d444218fbd7f59b2a4f86d933066c1db24/emailproxy.py#L562), if there is no access token, and then recurse (and exit early) if the token value changes after that reload.

Are there any cost implications to reloading from AWS Secrets? I can imagine an email client failing repeatedly might trigger a lot of requests...

Reply:
@simonrob the cost of AWS Secrets Manager is $0.40 per month + $0.05 per 10,000 API calls. So as long as the reload isn't happening in a rapid loop, I don't see a cost concern. If it happened once per minute for an entire 31-day month, it would cost US$0.22.

I'm not familiar with that section of code, so I'd defer to your judgement on the most sensible place to reload.

Reply:
Here's a quick go at a solution (completely untested, and unlikely to be the final approach) that is worth you trying out. Put this after [this line](https://github.com/michaelstepner/email-oauth2-proxy/blob/e11338d444218fbd7f59b2a4f86d933066c1db24/emailproxy.py#L562):

```python
        # try reloading from Secrets Manager to retrieve remotely-updated tokens if they exist 
        if not access_token and username in AppConfig._aws_secrets_accounts():
            AppConfig.reload()
            return OAuth2Helper.get_oauth2_credentials(username, password, recurse_retries=False)
```

Looking at this made me reconsider the `allow_catch_all_accounts` advanced configuration option. For consistency, having an `aws_secret` value in any parent account (e.g., the global [@], or, say [@gmail.com]) should use that secret to store values for all child accounts within that group.  I'll think about how is best to address this.

Reply:
Thanks @simonrob! Happy to test that out, although looking at my work week I'm not sure how soon I'll get to it this week.

Since we've completed the core functionality, do you think it would be worthwhile to merge this PR and split the reload functionality into a new PR? Or would you prefer to keep it in a single PR?

Reply:
> Looking at this made me reconsider the `allow_catch_all_accounts` advanced configuration option. For consistency, having an `aws_secret` value in any parent account (e.g., the global [@], or, say [@gmail.com]) should use that secret to store values for all child accounts within that group. I'll think about how is best to address this.

The simplest brute solution would be to only support `aws_secret` functionality for fully-specified accounts, and return a warning & ignore the `aws_secret` if it's specified for a catch_all account. Up to you how valuable it would be to get these two features to work together, but I don't foresee using that functionality personally.

Reply:
On second thought, I think the functionality already implemented so far will work fine for catch-all accounts (unless I'm mistaken). It's the reload step that is trickier.

Reply:
I think the catch-all functionality is quite valuable, but it should be relatively straightforward to incorporate that into the existing PR. I'll take a look when I get chance.

How did you get on with the reloading suggestion?

Reply:
I've had a busy week and haven't had a chance to test the reloading approach you suggested, but I look forward to testing it soon—perhaps this weekend.

Re: catch-all with AWS Secrets, I agree it will be best to integrate! I haven't used catch-all accounts before because I only use the proxy for a single account, but I suppose I can define a catch-all that applies to that single account in order to test.